KLR/NKI/Types.lean

```diff
@@ -0,0 +1,137 @@
+/-
+Copyright (c) 2025 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Mure
+-/
+import KLR.Core.Operators
+
+/-!
+# A type system for NKI
+
+The polymorphic type system for NKI consisting of two sorts: `SNat` and `STyp`.
+
+`SNat` represents the natural numbers used to denote tensor shapes.
+`STyp` is the universe of types that includes tensors parametrized by `SNat`s.
+
+All type parameters are universally quantified and must be given up front.
+We then use De Bruijn indices in the form of `Fin` type to index type parameter binders.
+-/
+
+namespace KLR.NKI
+
+/--
+A type universe for denoting tensor shapes.
+
+`nnat` is the total number of universal quantifiers in scope.
+
+Example:
+`∀ n m n, n` ≡ `(SNat.param 0 : SNat 3)`
+-/
+inductive SNat (nnat : Nat)
+  | param (idx : Fin nnat)
+  | const (n : Nat)
+
+/--
+The universe of types.
+
+`nnat` is the total number of universal quantifiers in scope for `SNat`s.
+
+`ntyp` is the total number of universal quantifiers in scope for `STyp`s.
+
+Example:
+`∀ n m n, ∀ t1 t2, tensor[t1, 10, m]` ≡ `(STyp.tensor [SNat.const 10, SNat.param 1] (STyp.param 1) : STyp 3 2)`
+-/
+inductive STyp (nnat ntyp : Nat)
+  | param (idx : Fin ntyp)
+  | none
+  | bool
+  | int
+  | float
+  | string
+  | tuple (typs : List (STyp nnat ntyp))
+  | tensor (shape : List (SNat nnat)) (dtype : Core.Dtype)
+  | func (dom ran : STyp nnat ntyp)
+
+namespace NominalTypes
+  /-!
+  # An alternative types definition using nominal variable binding
+
+  This requires a separate well-formedness proof to show that varaible references are valid.
+  -/
+
+  inductive Srt
```

Comment:
Just checking my understanding. We are distinguishing between Styp and Snat so we can keep tensor shape constants/variables apart from regular types for expressions, and vice versa. I had not thought of having different variables for these two things but makes sense.